paymaster/build_txn.go

```diff
@@ -0,0 +1,350 @@
+	// The Cairo version of the account contract (e.g. 1 or 2). Cairo 0 is not supported.
+	Version uint8 `json:"version"`
+}
+
+// Calls to be executed by the paymaster and the user account address that will be called
+type UserInvoke struct {
+	// The address of the user account
+	UserAddress *felt.Felt `json:"user_address"`
+	// The sequence of calls that the user wishes to perform
+	Calls []Call `json:"calls"`
+}
+
+// The object that defines an invocation of a function in a contract
+type Call struct {
+	// The address of the contract to invoke
+	To *felt.Felt `json:"to"`
+	// The selector of the function to invoke
+	Selector *felt.Felt `json:"selector"`
+	// The parameters passed to the function
+	Calldata []*felt.Felt `json:"calldata"`
+}
+
+// Execution parameters to be used when executing the transaction through the paymaster
+type UserParameters struct {
+	// Version of the execution parameters which is not tied to the 'execute from outside' version.
+	Version UserParamVersion `json:"version"`
+	// Fee mode to use for the execution
+	FeeMode FeeMode `json:"fee_mode"`
+	// Optional. Time constraint on the execution
+	TimeBounds *TimeBounds `json:"time_bounds"`
+}
+
+// An enum representing the version of the execution parameters
+type UserParamVersion string
+
+const (
+	// Represents the v1 of the execution parameters
+	UserParamV1 UserParamVersion = "0x1"
+)
+
+// MarshalJSON marshals the UserParamVersion to JSON.
+func (v UserParamVersion) MarshalJSON() ([]byte, error) {
+	return json.Marshal(string(v))
+}
+
+// UnmarshalJSON unmarshals the JSON data into a UserParamVersion.
+func (v *UserParamVersion) UnmarshalJSON(b []byte) error {
+	var s string
+	if err := json.Unmarshal(b, &s); err != nil {
+		return err
+	}
+	switch s {
+	case "0x1":
+		*v = UserParamV1
+	default:
+		return fmt.Errorf("invalid user parameter version: %s", s)
+	}
+
+	return nil
+}
+
+// An enum representing the fee mode to use for the transaction
+type FeeModeType string
+
+const (
+	// Specify that the transaction should be sponsored. This argument does not
+	// guaranteed sponsorship and will depend on the paymaster provider
+	FeeModeSponsored FeeModeType = "sponsored"
+	// Default fee mode where the transaction is paid by the user in the given gas token
+	FeeModeDefault FeeModeType = "default"
+)
+
+// MarshalJSON marshals the FeeModeType to JSON.
+func (feeMode FeeModeType) MarshalJSON() ([]byte, error) {
+	switch feeMode {
+	case FeeModeSponsored, FeeModeDefault:
+		return json.Marshal(string(feeMode))
+	}
+
+	return nil, fmt.Errorf("invalid fee mode: %s", feeMode)
+}
+
+// UnmarshalJSON unmarshals the JSON data into a FeeModeType.
+func (feeMode *FeeModeType) UnmarshalJSON(b []byte) error {
+	var s string
+	if err := json.Unmarshal(b, &s); err != nil {
+		return err
+	}
+
+	switch s {
+	case "sponsored":
+		*feeMode = FeeModeSponsored
+	case "default":
+		*feeMode = FeeModeDefault
+	default:
+		return fmt.Errorf("invalid fee mode: %s", s)
+	}
+
+	return nil
```

Comment:
All this string enums represent a lot of unnecessary memory usage (storing the string, when you only care about the tipe, hence you only need a unique id (using iotas)